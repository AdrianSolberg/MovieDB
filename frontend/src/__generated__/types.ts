export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  Date: { input: any; output: any; }
};

export type Movie = {
  _id: Scalars['Int']['output'];
  title: Scalars['String']['output'];
  vote_average: Scalars['Float']['output'];
  vote_count: Scalars['Int']['output'];
  status: Scalars['String']['output'];
  release_date: Scalars['Date']['output'];
  revenue: Scalars['Float']['output'];
  runtime: Scalars['Int']['output'];
  backdrop_path?: Maybe<Scalars['String']['output']>;
  budget: Scalars['Int']['output'];
  homepage?: Maybe<Scalars['String']['output']>;
  imdb_id?: Maybe<Scalars['String']['output']>;
  original_language: Scalars['String']['output'];
  original_title: Scalars['String']['output'];
  overview: Scalars['String']['output'];
  popularity: Scalars['Float']['output'];
  poster_path?: Maybe<Scalars['String']['output']>;
  tagline?: Maybe<Scalars['String']['output']>;
  genres: Array<Scalars['String']['output']>;
  production_companies: Array<Scalars['String']['output']>;
  production_countries: Array<Scalars['String']['output']>;
  spoken_languages: Array<Scalars['String']['output']>;
  keywords: Array<Scalars['String']['output']>;
};

export type Filters = {
  Genre: Array<Scalars['String']['output']>;
  Rating: Array<Scalars['String']['output']>;
  Decade: Array<Scalars['String']['output']>;
  Status: Array<Scalars['String']['output']>;
  Runtime: Array<Scalars['String']['output']>;
};

export type FiltersInput = {
  Genre: Array<Scalars['String']['input']>;
  Rating: Array<Scalars['String']['input']>;
  Decade: Array<Scalars['String']['input']>;
  Status: Array<Scalars['String']['input']>;
  Runtime: Array<Scalars['String']['input']>;
};

export enum SortingType {
  NewestFirst = 'NEWEST_FIRST',
  OldestFirst = 'OLDEST_FIRST',
  BestRated = 'BEST_RATED',
  WorstRated = 'WORST_RATED',
  LongestRuntime = 'LONGEST_RUNTIME',
  ShortestRuntime = 'SHORTEST_RUNTIME'
}

export type Query = {
  movie?: Maybe<Movie>;
  movies: Array<Movie>;
  filters?: Maybe<Filters>;
};


export type QueryMovieArgs = {
  id: Scalars['Int']['input'];
};


export type QueryMoviesArgs = {
  skip?: InputMaybe<Scalars['Int']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  filters?: InputMaybe<FiltersInput>;
  sortOption?: InputMaybe<SortingType>;
  search?: InputMaybe<Scalars['String']['input']>;
};

export type GetMovieQueryVariables = Exact<{
  movieId: Scalars['Int']['input'];
}>;


export type GetMovieQuery = { __typename?: 'Query', movie?: { __typename?: 'Movie', _id: number, title: string, vote_average: number, vote_count: number, status: string, release_date: any, revenue: number, runtime: number, backdrop_path?: string | null, budget: number, homepage?: string | null, imdb_id?: string | null, original_language: string, original_title: string, overview: string, popularity: number, poster_path?: string | null, tagline?: string | null, genres: Array<string>, production_companies: Array<string>, production_countries: Array<string>, spoken_languages: Array<string>, keywords: Array<string> } | null };

export type GetMoviesQueryVariables = Exact<{
  skip?: InputMaybe<Scalars['Int']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  filters?: InputMaybe<FiltersInput>;
  sortOption?: InputMaybe<SortingType>;
  search?: InputMaybe<Scalars['String']['input']>;
}>;


export type GetMoviesQuery = { __typename?: 'Query', movies: Array<{ __typename?: 'Movie', _id: number, title: string, vote_average: number, release_date: any, runtime: number, poster_path?: string | null }> };

export type GetFiltersQueryVariables = Exact<{ [key: string]: never; }>;


export type GetFiltersQuery = { __typename?: 'Query', filters?: { __typename?: 'Filters', Genre: Array<string>, Rating: Array<string>, Decade: Array<string>, Status: Array<string>, Runtime: Array<string> } | null };
