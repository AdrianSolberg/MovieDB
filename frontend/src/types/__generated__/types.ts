export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = {
    [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
    [SubKey in K]?: Maybe<T[SubKey]>;
};
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {
    [SubKey in K]: Maybe<T[SubKey]>;
};
export type MakeEmpty<
    T extends { [key: string]: unknown },
    K extends keyof T,
> = { [_ in K]?: never };
export type Incremental<T> =
    | T
    | {
          [P in keyof T]?: P extends " $fragmentName" | "__typename"
              ? T[P]
              : never;
      };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
    ID: { input: string; output: string };
    String: { input: string; output: string };
    Boolean: { input: boolean; output: boolean };
    Int: { input: number; output: number };
    Float: { input: number; output: number };
    DateTime: { input: Date; output: Date };
};

export type Movie = {
    __typename?: "Movie";
    _id: Scalars["Int"]["output"];
    title: Scalars["String"]["output"];
    vote_average: Scalars["Float"]["output"];
    vote_count: Scalars["Int"]["output"];
    status: Scalars["String"]["output"];
    release_date: Scalars["DateTime"]["output"];
    revenue: Scalars["Float"]["output"];
    runtime: Scalars["Int"]["output"];
    backdrop_path?: Maybe<Scalars["String"]["output"]>;
    budget: Scalars["Int"]["output"];
    homepage?: Maybe<Scalars["String"]["output"]>;
    imdb_id?: Maybe<Scalars["String"]["output"]>;
    original_language: Scalars["String"]["output"];
    original_title: Scalars["String"]["output"];
    overview: Scalars["String"]["output"];
    popularity: Scalars["Float"]["output"];
    poster_path?: Maybe<Scalars["String"]["output"]>;
    tagline?: Maybe<Scalars["String"]["output"]>;
    genres: Array<Scalars["String"]["output"]>;
    production_companies: Array<Scalars["String"]["output"]>;
    production_countries: Array<Scalars["String"]["output"]>;
    spoken_languages: Array<Scalars["String"]["output"]>;
    keywords: Array<Scalars["String"]["output"]>;
    reviews: Array<Review>;
};

export type Review = {
    __typename?: "Review";
    _id: Scalars["ID"]["output"];
    movie: Movie;
    username: Scalars["String"]["output"];
    rating: Scalars["Int"]["output"];
    comment: Scalars["String"]["output"];
    date: Scalars["DateTime"]["output"];
};

export type Filters = {
    __typename?: "Filters";
    Genre: Array<Scalars["String"]["output"]>;
    Rating: Array<Scalars["String"]["output"]>;
    Decade: Array<Scalars["String"]["output"]>;
    Status: Array<Scalars["String"]["output"]>;
    Runtime: Array<Scalars["String"]["output"]>;
};

export type FiltersInput = {
    Genre: Array<Scalars["String"]["input"]>;
    Rating: Array<Scalars["String"]["input"]>;
    Decade: Array<Scalars["String"]["input"]>;
    Status: Array<Scalars["String"]["input"]>;
    Runtime: Array<Scalars["String"]["input"]>;
};

export enum SortingType {
    MOST_POPULAR = "MOST_POPULAR",
    NEWEST_FIRST = "NEWEST_FIRST",
    OLDEST_FIRST = "OLDEST_FIRST",
    BEST_RATED = "BEST_RATED",
    WORST_RATED = "WORST_RATED",
    LONGEST_RUNTIME = "LONGEST_RUNTIME",
    SHORTEST_RUNTIME = "SHORTEST_RUNTIME",
}

export type Query = {
    __typename?: "Query";
    movie?: Maybe<Movie>;
    movies: Array<Movie>;
    filters: Filters;
    latestReviews: Array<Review>;
    userReviews: Array<Review>;
};

export type QuerymovieArgs = {
    id: Scalars["Int"]["input"];
};

export type QuerymoviesArgs = {
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
    filters?: InputMaybe<FiltersInput>;
    sortOption?: InputMaybe<SortingType>;
    search?: InputMaybe<Scalars["String"]["input"]>;
};

export type QuerylatestReviewsArgs = {
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
};

export type QueryuserReviewsArgs = {
    username: Scalars["String"]["input"];
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
};

export type Mutation = {
    __typename?: "Mutation";
    addReview: Movie;
    deleteReview: Movie;
};

export type MutationaddReviewArgs = {
    movie_id: Scalars["Int"]["input"];
    username: Scalars["String"]["input"];
    rating: Scalars["Int"]["input"];
    comment: Scalars["String"]["input"];
};

export type MutationdeleteReviewArgs = {
    _id: Scalars["ID"]["input"];
};

export type GetMoviesQueryVariables = Exact<{
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
    filters?: InputMaybe<FiltersInput>;
    sortOption?: InputMaybe<SortingType>;
    search?: InputMaybe<Scalars["String"]["input"]>;
}>;

export type GetMoviesQuery = {
    __typename?: "Query";
    movies: Array<{
        __typename?: "Movie";
        _id: number;
        title: string;
        vote_average: number;
        release_date: Date;
        runtime: number;
        poster_path?: string | null;
    }>;
};

export type GetMovieQueryVariables = Exact<{
    movieId: Scalars["Int"]["input"];
}>;

export type GetMovieQuery = {
    __typename?: "Query";
    movie?: {
        __typename?: "Movie";
        _id: number;
        title: string;
        vote_average: number;
        vote_count: number;
        status: string;
        release_date: Date;
        revenue: number;
        runtime: number;
        backdrop_path?: string | null;
        budget: number;
        homepage?: string | null;
        imdb_id?: string | null;
        original_language: string;
        original_title: string;
        overview: string;
        popularity: number;
        poster_path?: string | null;
        tagline?: string | null;
        genres: Array<string>;
        production_companies: Array<string>;
        production_countries: Array<string>;
        spoken_languages: Array<string>;
        keywords: Array<string>;
        reviews: Array<{
            __typename?: "Review";
            _id: string;
            username: string;
            rating: number;
            comment: string;
            date: Date;
        }>;
    } | null;
};

export type GetFiltersQueryVariables = Exact<{ [key: string]: never }>;

export type GetFiltersQuery = {
    __typename?: "Query";
    filters: {
        __typename?: "Filters";
        Genre: Array<string>;
        Rating: Array<string>;
        Decade: Array<string>;
        Status: Array<string>;
        Runtime: Array<string>;
    };
};

export type AddReviewMutationVariables = Exact<{
    movieId: Scalars["Int"]["input"];
    username: Scalars["String"]["input"];
    rating: Scalars["Int"]["input"];
    comment: Scalars["String"]["input"];
}>;

export type AddReviewMutation = {
    __typename?: "Mutation";
    addReview: {
        __typename?: "Movie";
        reviews: Array<{
            __typename?: "Review";
            _id: string;
            username: string;
            rating: number;
            comment: string;
            date: Date;
        }>;
    };
};
