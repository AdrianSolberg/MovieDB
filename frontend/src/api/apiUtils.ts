import { Review } from "@/types/__generated__/types";

type ResponseReview = {
    __typename?: "Review";
    _id: string;
    username: string;
    rating: number;
    comment: string;
    date: Date;
};

/**
 *  Make sure that the date-field is correctly initialized as a Date-object
 *  when getting reviews as a response from a query.
 */
export function parseReviewResponse(inputReviews: ResponseReview[]): Review[] {
    return inputReviews.map((review) => ({
        ...review,
        date: new Date(review.date),
    })) as Review[];
}
