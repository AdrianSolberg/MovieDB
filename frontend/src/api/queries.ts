import { gql } from "@/types/__generated__";

export const GET_MOVIES = gql(`
    query GetMovies(
        $skip: Int
        $limit: Int
        $filters: FiltersInput
        $sortOption: SortingType
        $search: String
    ) {
        movies(
            skip: $skip
            limit: $limit
            filters: $filters
            sortOption: $sortOption
            search: $search
        ) {
            _id
            title
            vote_average
            release_date
            runtime
            poster_path
        }
    }
`);

export const GET_MOVIE = gql(`
    query GetMovie($movieId: Int!) {
        movie(id: $movieId) {
            _id
            title
            vote_average
            vote_count
            status
            release_date
            revenue
            runtime
            backdrop_path
            budget
            homepage
            imdb_id
            original_language
            original_title
            overview
            popularity
            poster_path
            tagline
            genres
            production_companies
            production_countries
            spoken_languages
            keywords
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);

export const GET_FILTERS = gql(`
    query GetFilters {
        filters {
            Genre
            Rating
            Decade
            Status
            Runtime
        }
    }
`);

export const GET_LATEST_REVIEWS = gql(`
    query GetLatestReviews(
        $skip: Int,
        $limit: Int
    ) {
        latestReviews(
            skip: $skip,
            limit: $limit
        ) {
            _id
            movie {
                _id
                title
                poster_path
            }
            username
            rating
            comment
            date
        }
    }
`);

export const GET_USER_REVIEWS = gql(`
    query GetUserReviews(
        $username: String!,
        $skip: Int,
        $limit: Int
    ) {
        userReviews(
            username: $username,
            skip: $skip,
            limit: $limit
        ) {
            _id
            movie {
                _id
                title
                poster_path
            }
            rating
            comment
            date
        }
    }
`);

export const ADD_REVIEW = gql(`
    mutation AddReview(
        $movieId: Int!,
        $username: String!,
        $rating: Int!,
        $comment: String!
    ) {
        addReview(
            movie_id: $movieId,
            username: $username,
            rating: $rating,
            comment: $comment
        ) {
            _id
            vote_average
            vote_count
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);

export const DELETE_REVIEW = gql(`
    mutation DeleteReview($id: ID!) {
        deleteReview(_id: $id) {
            _id
            vote_average
            vote_count
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);
