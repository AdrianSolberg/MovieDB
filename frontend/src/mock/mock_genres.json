[
    {
        "id": 0,
        "name": "Drama"
    },
    {
        "id": 1,
        "name": "Romance"
    },
    {
        "id": 2,
        "name": "Western"
    },
    {
        "id": 3,
        "name": "Animation"
    },
    {
        "id": 4,
        "name": "War"
    },
    {
        "id": 5,
        "name": "Science Fiction"
    },
    {
        "id": 6,
        "name": "Comedy"
    },
    {
        "id": 7,
        "name": "Fantasy"
    },
    {
        "id": 8,
        "name": "Adventure"
    },
    {
        "id": 9,
        "name": "Crime"
    },
    {
        "id": 10,
        "name": "Thriller"
    },
    {
        "id": 11,
        "name": "Family"
    },
    {
        "id": 12,
        "name": "Action"
    },
    {
        "id": 13,
        "name": "Mystery"
    }
]
