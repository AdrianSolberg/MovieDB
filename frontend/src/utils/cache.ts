import { makeVar } from "@apollo/client";
import { Filters, SortingType } from "@/types/movieTypes";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);


const storedFilters = sessionStorage.getItem("filters");
const initialFilters: Filters = storedFilters ? JSON.parse(storedFilters) : {
    Genre: [],
    Rating: [],
    Decade: [],
    Status: [],
    Runtime: [],
};
export const filtersVar = makeVar<Filters>(initialFilters);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);

export const usernameVar = makeVar<string>(
    localStorage.getItem("username") ?? "Guest"
);
