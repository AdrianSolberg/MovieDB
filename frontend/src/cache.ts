import { makeVar } from "@apollo/client";
import { SortingType } from "./utils/searchSortAndFilter";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);

export const filtersVar = makeVar<{ [key: string]: string[] }>(
    JSON.parse(sessionStorage.getItem("filters") ?? "{}")
);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);
