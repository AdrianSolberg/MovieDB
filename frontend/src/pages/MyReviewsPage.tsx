import ReviewCard from "@/components/ReviewCard";
import { Button } from "@/shadcn/components/ui/button";
import { useMemo, useState } from "react";
import { Link } from "react-router-dom";
import { useQuery, useReactiveVar } from "@apollo/client";
import { GET_USER_REVIEWS } from "../api/queries";
import { Review } from "@/types/__generated__/types";
import { usernameVar } from "@/utils/cache";

const MyReviewsPage = () => {
    const username = useReactiveVar(usernameVar);
    const [isMoreReviews, setIsMoreReviews] = useState<boolean>(true);
    const LIMIT = 20;

    const { data, loading, error, fetchMore } = useQuery(GET_USER_REVIEWS, {
        variables: {
            username: username,
            skip: 0,
            limit: LIMIT,
        },
        onCompleted: (data) => {
            if (data.userReviews) {
                setIsMoreReviews(
                    data.userReviews.length >= LIMIT &&
                        data.userReviews.length % LIMIT === 0
                );
            }
        },
    });

    const userReviews = useMemo(
        () => data?.userReviews as Review[] | undefined,
        [data]
    );

    const handleLoadMore = () => {
        fetchMore({ variables: { skip: userReviews?.length } }).then(
            (fetchMoreResult) => {
                setIsMoreReviews(
                    fetchMoreResult.data.userReviews.length === LIMIT
                );
            }
        );
    };

    if (loading) {
        return (
            <section className="mt-6 w-dvw text-center">
                <h1 className="text-2xl">Loading...</h1>
            </section>
        );
    }

    if (error || !userReviews) {
        return (
            <section className="mt-6 w-dvw text-center">
                <h1 className="text-2xl">Something went wrong!</h1>
                <p className="text-primary">Try to refresh</p>
            </section>
        );
    }

    if (userReviews.length === 0) {
        return (
            <section className="mt-6 text-center">
                <h1 className="text-2xl">No reviews found</h1>
                <Link to="/" className="text-primary hover:underline">
                    Return to home page
                </Link>
            </section>
        );
    }

    return (
        <main className="mx-auto mt-8 max-w-6xl px-4 pb-4">
            <h1 className="mb-6 text-center text-3xl font-bold">My Reviews</h1>
            <ul className="space-y-6">
                {userReviews?.map((review) => (
                    <li key={review._id}>
                        <ReviewCard
                            review={{
                                ...review,
                                username,
                            }}
                        />
                    </li>
                ))}
            </ul>
            {isMoreReviews && (
                <div className="flex justify-center">
                    <Button
                        size="lg"
                        className="m-10"
                        onClick={handleLoadMore}
                        disabled={loading}
                    >
                        {loading ? "Loading..." : "Load More"}
                    </Button>
                </div>
            )}
        </main>
    );
};

export default MyReviewsPage;
