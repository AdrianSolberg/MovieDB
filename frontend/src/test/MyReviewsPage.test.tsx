import { GET_USER_REVIEWS } from "@/api/queries";
import { all_reviews } from "./mock/util";
import "@testing-library/jest-dom";
import { usernameVar } from "@/utils/cache";
import { MemoryRouter } from "react-router-dom";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { act, render, screen, waitFor } from "@testing-library/react";
import MyReviewsPage from "@/pages/MyReviewsPage";
import { vi } from "vitest";
import { Review } from "@/types/__generated__/types";
import { useReactiveVar } from "@apollo/client";

const mockUserReviews = [
    {
        request: {
            query: GET_USER_REVIEWS,
            variables: {
                username: "other_user",
                skip: 0,
                limit: 20,
            },
        },
        result: {
            data: {
                userReviews: [],
            },
        },
    },
    {
        request: {
            query: GET_USER_REVIEWS,
            variables: {
                username: "test_user",
                skip: 0,
                limit: 20,
            },
        },
        result: {
            data: {
                userReviews: all_reviews.slice(0, 20),
            },
        },
    },
];

const mockUserReviewsError = [
    {
        request: {
            query: GET_USER_REVIEWS,
            variables: {
                username: "test_user",
                skip: 0,
                limit: 20,
            },
        },
        error: new Error("Error fetching movie"),
    },
];

vi.mock("../components/ReviewCard", () => ({
    default: vi.fn(({ review }: { review: Review }) => (
        <div data-testid="review-card">
            <span data-testid="review-username">{review.username}</span>
            <span data-testid="review-rating">{review.rating}</span>
            <span data-testid="review-comment">{review.comment}</span>
        </div>
    )),
}));

vi.mock("../components/Loader", () => ({
    default: vi.fn(() => <div data-testid="loader" />),
}));

vi.mock("@/utils/cache", () => ({
    usernameVar: vi.fn(),
}));

vi.mock("@apollo/client", async () => {
    const original = await vi.importActual("@apollo/client");
    return {
        ...original,
        useReactiveVar: vi.fn(),
    };
});

const mockNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
    const actualModule = await vi.importActual("react-router-dom");
    return {
        ...actualModule,
        useNavigate: () => mockNavigate,
    };
});

describe("ActivityPage", () => {
    const renderComponent = (mocks: MockedResponse[] | undefined) => {
        return render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <MemoryRouter>
                    <MyReviewsPage />
                </MemoryRouter>
            </MockedProvider>
        );
    };

    beforeEach(() => {
        vi.mocked(usernameVar).mockReturnValue("test_user");
        vi.mocked(useReactiveVar).mockImplementation((varFn) => varFn());
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    it("matches snapshot when loading", () => {
        const { asFragment } = renderComponent(mockUserReviews);
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot when user reviews are loaded", async () => {
        const { asFragment } = renderComponent(mockUserReviews);
        await screen.findByText("My Reviews");
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot when user has no reviews", async () => {
        vi.mocked(usernameVar).mockReturnValue("other_user");
        const { asFragment } = renderComponent(mockUserReviews);
        await screen.findByText("You have not added any reviews yet");
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot on error", async () => {
        const { asFragment } = renderComponent(mockUserReviewsError);
        await screen.findByText("Something went wrong!");
        expect(asFragment()).toMatchSnapshot();
    });

    it("displays 'You have not added any reviews yet' on empty reviews", async () => {
        vi.mocked(usernameVar).mockReturnValue("other_user");
        renderComponent(mockUserReviews);
        await waitFor(() =>
            expect(
                screen.getByText("You have not added any reviews yet")
            ).toBeInTheDocument()
        );
    });

    it("displays 'Something went wrong!' on error", async () => {
        renderComponent(mockUserReviewsError);
        await waitFor(() =>
            expect(
                screen.getByText("Something went wrong!")
            ).toBeInTheDocument()
        );
    });

    it("renders 20 reviews", async () => {
        renderComponent(mockUserReviews);
        //All reviews have username 'test_user'
        await waitFor(() =>
            expect(screen.getAllByText("test_user")).toHaveLength(20)
        );
    });

    it("redirects to homepage on logout", async () => {
        mockNavigate.mockClear();
        renderComponent(mockUserReviews);
        act(() => vi.mocked(usernameVar).mockReturnValue("Guest"));
        await waitFor(() => expect(mockNavigate).toHaveBeenCalledWith("/"));
    });

    it("does not redirect to homepage on user change", async () => {
        mockNavigate.mockClear();
        renderComponent(mockUserReviews);
        act(() => vi.mocked(usernameVar).mockReturnValue("other_user"));
        await waitFor(() => expect(mockNavigate).not.toHaveBeenCalled());
    });
});
