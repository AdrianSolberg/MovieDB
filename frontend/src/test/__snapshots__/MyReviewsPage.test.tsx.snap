// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyReviewsPage > matches snapshot on error 1`] = `
<DocumentFragment>
  <section
    class="mt-6 w-dvw text-center"
  >
    <h1
      class="text-2xl"
    >
      Something went wrong!
    </h1>
    <p
      class="text-primary"
    >
      Try to refresh
    </p>
  </section>
</DocumentFragment>
`;

exports[`MyReviewsPage > matches snapshot when loading 1`] = `
<DocumentFragment>
  <section
    class="mt-6 w-dvw text-center"
  >
    <div
      data-testid="loader"
    />
  </section>
</DocumentFragment>
`;

exports[`MyReviewsPage > matches snapshot when user has no reviews 1`] = `
<DocumentFragment>
  <section
    class="mt-6 text-center"
  >
    <h1
      class="text-2xl"
    >
      You have not added any reviews yet
    </h1>
    <a
      class="text-primary hover:underline"
      href="/"
    >
      Return to home page
    </a>
  </section>
</DocumentFragment>
`;

exports[`MyReviewsPage > matches snapshot when user reviews are loaded 1`] = `
<DocumentFragment>
  <section
    class="mx-auto mt-8 max-w-6xl px-4 pb-4"
  >
    <h1
      class="mb-6 text-center text-3xl font-bold"
    >
      My Reviews
    </h1>
    <div>
      <ul
        class="space-y-6"
      >
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              4
            </span>
            <span
              data-testid="review-comment"
            >
              good movie!
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              22
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              21
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              20
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              19
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              18
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              4
            </span>
            <span
              data-testid="review-comment"
            >
              17
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              16
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              15
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              14
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              13
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              12
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              11
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              10
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              9
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              8
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              7
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              5
            </span>
            <span
              data-testid="review-comment"
            >
              6
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              4
            </span>
            <span
              data-testid="review-comment"
            >
              5
            </span>
          </div>
        </li>
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              test_user
            </span>
            <span
              data-testid="review-rating"
            >
              3
            </span>
            <span
              data-testid="review-comment"
            >
              4
            </span>
          </div>
        </li>
      </ul>
      <div
        data-testid="loader"
      />
    </div>
  </section>
</DocumentFragment>
`;
